PLG-2665 Added the CLI tutorial (#2984)
harness/developer-hub: tutorials/cd-pipelines/kubernetes/manifest.md

:::

## Building the Pipeline Resources 

1. **Create an API key**, make sure to copy and save it for future use while creating resources for the pipeline. 

> 2. **Download the Harness CLI**

Reviewer: ```suggestion
2. **Download the Harness CLI**.
```

```mdx-code-block
<Tabs>
<TabItem value="MacOS">
```
```
curl -L0 https://github.com/harness/harness-cli/releases/download/v0.0.10-alpha/harness-v0.0.10-alpha-darwin-amd64.tar.gz
```
```mdx-code-block
</TabItem>
<TabItem value="Linux">
```
```
curl -L0 https://github.com/harness/harness-cli/releases/download/v0.0.10-alpha/harness-v0.0.10-alpha-linux-amd64.tar.gz
```
```mdx-code-block
</TabItem>
<TabItem value="Windows">
```
```
curl -L0 https://github.com/harness/harness-cli/releases/download/v0.0.10-alpha/harness-v0.0.10-alpha-windows-amd64.zip
```
```mdx-code-block
</TabItem>
</Tabs>
```

> 3. Add github username and PAT you generated above under **Enter Your Github Details**.

Reviewer: ```suggestion
3. Add the GitHub username and PAT you generated above under **Enter Your Github Details**.
```

4. **Pipeline Setup**

> Under this step, you would see a bunch of CLI command to run, before procceding with that, please follow the steps below to create the pipeline resources. 

Reviewer: ```suggestion
Under this step, you would see a bunch of CLI commands to run. Before proceeding with that, please follow the steps below to create the pipeline resources. 
```

> - Use the follwwing command to clone the repo and get into the root folder conating the YAMLS used in this tutorial. 

Reviewer: ```suggestion
- Use the following command to clone the repo and get into the root folder containing the YAMLS used in this tutorial. 
```

    ```
    git clone https://github.com/harness-community/harnesscd-example-apps.git 
    cd harness-example-apps/guestbook/harnesscd-pipeline
    ```
- **Login Using CLI**
    - For this step we will use **account ID** from the Harness URL and the **x-api-key** created above. 
    - Example of Harness URL: `https://app.harness.io/ng/account/<account ID>/cd/orgs/<org ID>/projects/<Project ID>/`

>     - Now use the following CLI command to login. 

Reviewer: ```suggestion
    - Now use the following CLI command to log in. 
```

    ```
    harness login --api-key <YOUR API KEY> --account-id <YOUR ACCOUNT ID>
    ```
    #### Secrets
    <details open>
    <summary>What are Harness secrets?</summary>
    Harness offers built-in secret management for encrypted storage of sensitive information. Secrets are decrypted when needed, and only the private network-connected Harness delegate has access to the key management system. You can also integrate your own secret manager. To learn more about secrets in Harness, go to [Harness Secret Manager Overview](https://developer.harness.io/docs/platform/Secrets/Secrets-Management/harness-secret-manager-overview/).
    </details>
    

>     - Use the follwoing command to add Github PAT created above for secret.

Reviewer: ```suggestion
    - Use the following command to add the GitHub PAT you created previously for your secret.
```

    
    ```
    harness secret --token <YOUR GITHUB PAT>
    ```
    #### Connectors
    <details open>
    <summary>What are connectors?</summary>
    Connectors in Harness enable integration with 3rd party tools, providing authentication and operations during pipeline runtime. For instance, a GitHub connector facilitates authentication and fetching files from a GitHub repository within pipeline stages. Explore connector how-tos [here](https://developer.harness.io/docs/category/connectors).
    </details>
    - Replace **GITHUB_USERNAME** with your GitHub account username in the `github-connector.yaml` 
    - In `projectIdentifier`, verify that the project identifier is correct. You can see the Id in the browser URL (after `account`). If it is incorrect, the Harness YAML editor will suggest the correct Id.

>     - Now create the **GitHub connector** using the following CLI Command

Reviewer: ```suggestion
    - Now create the **GitHub connector** using the following CLI command:
```

        ```
        harness connector --file github-connector.yml apply --git-user <YOUR GITHUB USERNAME>
        ```
    - Please check the delegate name to be `helm-delegate` in the `kubernetes-connector.yml`

>     - Create the **Kubernetes connector** using the following CLI Command

Reviewer: ```suggestion
    - Create the **Kubernetes connector** using the following CLI command:
```

        
        ```
        harness connector --file kubernetes-connector.yml apply --delegate-name kubernetes-delegate
        ```
### Environment

<details open>
<summary>What are Harness environments?</summary>
Environments define the deployment location, categorized as **Production** or **Pre-Production**. Each environment includes infrastructure definitions for VMs, Kubernetes clusters, or other target infrastructures. To learn more about environments, go to [Environments overview](https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/environments/environment-overview/).
</details>
    

>  - Use the follwing CLI Command to create **Environments** in your Harness project

Reviewer: ```suggestion
 - Use the follwing CLI Command to create **Environments** in your Harness project:
```

        
        ```
        harness environment --file environment.yml apply
        ```
        

> - In your new environment, add **Infrastructure Definitions** using the following CLI command.

Reviewer: ```suggestion
- In your new environment, add **Infrastructure Definitions** using the following CLI command:
```

```

Finally, it's time to execute your pipeline. 

> 1. Go to the pipeline tab on left nav bar and you can see your pipeline created.

Reviewer: ```suggestion
1. Go to the pipeline tab on left navigation bar and you can see your pipeline created.
```
